pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1065,42 +976,55 @@ type httpListenerOpts struct {
 
 // filterChainOpts describes a filter chain: a set of filters with the same TLS context
 type filterChainOpts struct {
-	filterChainName  string
-	sniHosts         []string
-	destinationCIDRs []string
-	metadata         *core.Metadata
-	tlsContext       *auth.DownstreamTlsContext
-	httpOpts         *httpListenerOpts
-	match            *listener.FilterChainMatch
-	listenerFilters  []*listener.ListenerFilter
-	networkFilters   []*listener.Filter
-	filterChain      istionetworking.FilterChain
+	// Matching criteria. Will eventually turn into FilterChainMatch
+	sniHosts             []string
+	destinationCIDRs     []string
+	applicationProtocols []string
+	transportProtocol    string
+
+	// Arbitrary metadata to attach to the filter
+	metadata *core.Metadata
+
+	// TLS configuration for the filter
+	tlsContext *auth.DownstreamTlsContext
+
+	// Set if this is for HTTP. Cannot be set with networkFilters
+	httpOpts *httpListenerOpts
+	// Set if this is for TCP chain. Cannot be set with httpOpts
+	networkFilters []*listener.Filter
 }
 
-// buildListenerOpts are the options required to build a Listener
-type buildListenerOpts struct {
-	// nolint: maligned
-	push              *model.PushContext
-	proxy             *model.Proxy
-	bind              string
-	extraBind         []string
+// gatewayListenerOpts are the options required to build a gateway Listener
+type gatewayListenerOpts struct {
+	push  *model.PushContext
+	proxy *model.Proxy
+
+	bindToPort bool
+	bind       string
+	extraBind  []string
+
 	port              *model.Port
 	filterChainOpts   []*filterChainOpts
-	bindToPort        bool
-	skipUserFilters   bool
-	needHTTPInspector bool
 	needPROXYProtocol bool
-	class             istionetworking.ListenerClass
-	service           *model.Service
-	transport         istionetworking.TransportProtocol
 }
 
-// buildListener builds and initializes a Listener proto based on the provided opts. It does not set any filters.
+// outboundListenerOpts are the options to build an outbound listener
+type outboundListenerOpts struct {
+	push  *model.PushContext
+	proxy *model.Proxy
+
+	bind listenerBinding
+	cidr string
+
+	port    *model.Port
+	service *model.Service
+}
+
+// buildGatewayListener builds and initializes a Listener proto based on the provided opts. It does not set any filters.
 // Optionally for HTTP filters with TLS enabled, HTTP/3 can be supported by generating QUIC Mirror filters for the
 // same port (it is fine as QUIC uses UDP)
-func buildListener(opts buildListenerOpts, trafficDirection core.TrafficDirection) *listener.Listener {
+func buildGatewayListener(opts gatewayListenerOpts, transport istionetworking.TransportProtocol) *listener.Listener {
```

Comment:
I did not compare closely with `buildListenerFromEntry`, since we separate gateway and outbound listener built, will there be any diverge later

Reply:
Only if we intend to. Gateway and outbound listener building is not very similar actually, so most of the logic is just `if gateway { ... }`